File: mobile/ios/CortexV2-Emotiv-iOS/CortexV2-Emotiv-ObjC/CortexV2-Emotiv-ObjC/DataStreamViewController.m
//
//  DataStreamViewController.m
//  CortexV2-Emotiv-ObjC
//
//  Created by nvtu on 2/24/20.
//  Copyright © 2020 Emotiv. All rights reserved.
//

#import "DataStreamViewController.h"
#import "Config.h"

@interface DataStreamViewController ()

@end

@implementation DataStreamViewController

- (id) initActivateSession: (BOOL) activateSession license: (NSString *) license stream: (NSString *) stream {
    self = [super init];
    if (self) {
        self->activateSession = activateSession;
        self->license = license;
        self->stream = stream;
    }
    return self;
}

- (void) viewDidLoad {
    [super viewDidLoad];
    
    self.headsetTableView.dataSource = self;
    self.headsetTableView.delegate = self;
    [self.headsetTableView registerNib:[UINib nibWithNibName:@"HeadsetTableViewCell" bundle:nil] forCellReuseIdentifier:@"cellID"];
    headsetList = [[NSArray alloc] init];
    client = [CortexClient shareInstance];
    [self start];
    [self handleCortexEvent];
    [self handleHeadsetEvent];
    
    
    // Do any additional setup after loading the view.
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    [self handleUI];
}

- (void) start {
    [client open];
}

- (IBAction)getUserLogin:(id)sender {
    [client getUserLogin];
}

- (IBAction)hasAccessRight:(id)sender {
    [client hasAccessRight:[Config getClientID] secret:[Config getClientSecret]];
}

- (IBAction)requestAccess:(id)sender {
    [client requestAccess:[Config getClientID] secret:[Config getClientSecret]];
}

- (IBAction)authorize:(id)sender {
    [client authorize:[Config getClientID] secret:[Config getClientSecret] license:self->license debit:self->activateSession ? 1 : 0];
}

- (IBAction)queryHeadset:(id)sender {
    [finder findHeadsets];
}

- (IBAction)createSession:(id)sender {
    [client createSession:self->token headsetId:self->headset->headsetId activate:self->activateSession];
}

- (IBAction)createRecord:(id)sender {
    [client createRecord:self->token sessionId:self->sessionId title:@"Cortex Examples Swift"];
}

- (IBAction)injectMarker:(id)sender {
    NSTimeInterval milisecondedDate = ([[NSDate date] timeIntervalSince1970] * 1000);
    [client injectMarker:self->token sessionId:self->sessionId label:@"test1" value:41 time:(int64_t)milisecondedDate];
}

- (IBAction)stopRecord:(id)sender {
    [client stopRecord:self->token sessionId:self->sessionId];
}

- (IBAction)openTrainingView:(id)sender {
    if (!self->sessionId.length) {
        NSLog(@"Please create session");
    } else {
        if ([self->stream isEqual:@"session"]) {
            [client closeSession:self->token sessionId:self->sessionId];
        } else {
            [self->_dataView setHidden:true];
            [self->_trainingView setHidden:false];
        }
    }
}

- (IBAction)subcribe:(id)sender {
    [client subscribe:self->token sessionId:self->sessionId stream:self->stream];
}

- (IBAction)unsubcribe:(id)sender {
    [client unsubscribe:self->token sessionId:self->sessionId stream:self->stream];
}

- (IBAction)getLicenseInfo:(id)sender {
    [client getLicenseInfos:self->token];
}

- (IBAction)getUserInfo:(id)sender {
    [client getUserInformation:self->token];
}

- (IBAction)subscribeTraining:(id)sender {
    [client subscribe:self->token sessionId:self->sessionId stream:@"sys"];
}

- (IBAction)loadProfile:(id)sender {
    [client loadProfile:self->token headsetId:self->headset->headsetId profileName:[Config getTrainingProfileName]];
}

- (IBAction)createProfile:(id)sender {
    [client createProfile:self->token profileName:[Config getTrainingProfileName]];
}

- (IBAction)clickTraining1:(id)sender {
    self->action = @"neutral";
    [client training:self->token sessionId:self->sessionId detection:self->stream action:self->action control:@"start"];
    [self enableButton:false];
    NSLog(@"waiting training result from Cortex");
}

- (IBAction)clickTraining2:(id)sender {
    self->action = [self->stream isEqual:@"mentalCommand"] ? @"push" : @"smile";
    [client training:self->token sessionId:self->sessionId detection:self->stream action:self->action control:@"start"];
    [self enableButton:false];
    NSLog(@"waiting training result from Cortex");
}

- (void) handleUI {
    NSString *path = [[NSBundle mainBundle] pathForResource:@"Info" ofType:@"plist"];
    if (path != nil) {
        NSDictionary * nsDictionary = [[NSDictionary alloc] initWithContentsOfFile:path];
        NSString *stream = nsDictionary[@"STREAM_NAME"];
        
        if (stream != nil) {
            BOOL activateSession = false;
            NSString *license = @"";
            if ([stream  isEqual: @"eeg"] || [stream  isEqual: @"marker"]) {
                NSLog(@"#");
                NSLog(@"#####");
                NSLog(@"Reminder: to subscribe to the EEG data stream, you must get an appropriate licence from Emotiv.");
                NSLog(@"#####");
                NSLog(@"#");
                license = @""; // you can put your license id here
                activateSession = true;
            }
            self->stream = stream;
            self->license = license;
            self->activateSession = activateSession;
        }
        
    }
    NSString *nameStream = @"";
    if ([stream isEqual:@"mot"]) {
        nameStream = @"motion";
    } else if ([stream isEqual:@"eeg"]) {
        nameStream = @"eeg";
    } else if ([stream isEqual:@"pow"]) {
        nameStream = @"Band Power";
    } else if ([stream isEqual:@"com"]) {
        nameStream = @"MC";
    } else if ([stream isEqual:@"fac"]) {
        nameStream = @"FE";
    } else if ([stream isEqual:@"met"]) {
        nameStream = @"PM";
    } else if ([stream isEqual:@"facialExpression"]) {
        [_openTrainingViewBtn setHidden:false];
        [_subcribeBtn setHidden:true];
        [_unsubscribeBtn setHidden:true];
        [_training1Btn setTitle:@"train FE Neutral" forState:UIControlStateNormal];
        [_training2Btn setTitle:@"train FE Smile" forState:UIControlStateNormal];
    } else if ([stream isEqual:@"mentalCommand"]) {
        [_openTrainingViewBtn setHidden:false];
        [_subcribeBtn setHidden:true];
        [_unsubscribeBtn setHidden:true];
        [_training1Btn setTitle:@"train MC Neutral" forState:UIControlStateNormal];
        [_training2Btn setTitle:@"train MC Push" forState:UIControlStateNormal];
    } else if ([stream isEqual:@"marker"]) {
        [_subscribeView setHidden:true];
        [_markerView setHidden:false];
    }
    
    if(![stream isEqual:@"authorize"]) {
        [_authorizeView setHidden:true];
        [_dataView setHidden:false];
    }
    
    if([stream isEqual:@"session"]) {
        [_openTrainingViewBtn setHidden:false];
        [_subcribeBtn setHidden:true];
        [_unsubscribeBtn setHidden:true];
        [_openTrainingViewBtn setTitle:@"updateSession(Close)" forState:UIControlStateNormal];
    }

    [_subcribeBtn setTitle:[NSString stringWithFormat:@"%@ %@", @"subcribe ", nameStream] forState:UIControlStateNormal];
    [_unsubscribeBtn setTitle:[NSString stringWithFormat:@"%@ %@", @"unsubcribe ", nameStream] forState:UIControlStateNormal];
    [NSTimer scheduledTimerWithTimeInterval: 60.0 repeats: false block:^(NSTimer * timer) {
        [self.warningText setHidden:true];
    }];
}

- (void) handleCortexEvent {
    __unsafe_unretained typeof(self) weakSelf = self;

    client.onConnected = ^(){
        NSLog(@"Connected.");
        if (![weakSelf->stream isEqual:@"authorize"]) {
            [weakSelf->client authorize:[Config getClientID] secret:[Config getClientSecret] license:weakSelf->license debit:weakSelf->activateSession ? 1 : 0];
        }
    };
    
    client.onDisconnected = ^(){
        NSLog(@"Disconnected.");
    };
    
    client.onErrorReceived = ^(NSString *method, int code, NSString *message) {
    };
    
    client.onGetUserLoginOk = ^(NSString * emotivId) {
        if ([emotivId isEqual: @""]) {
            NSLog(@"First, you must login with EMOTIV App");
            return;
        }
        NSLog(@"You are logged in with the EmotivId %@", emotivId);
    };
    
    client.onUnsubscribeOk = ^(NSArray * streams) {
        [weakSelf->client closeSession: weakSelf->token sessionId: weakSelf->sessionId];
    };
    
    client.onCloseSessionOk = ^(){
        NSLog(@"Session closed.");
    };
    
    client.onAuthorizeOk = ^(NSString *token) {
        weakSelf->token = token;
        NSLog(@"Authorize successful, token: %@",token);
        // next step: open a session for the headset
        if (![weakSelf->stream isEqual:@"authorize"]) {
            [weakSelf->finder findHeadsets];
        }
    };
    
    client.onSubscribeOk = ^(NSArray * streams) {
        NSLog(@"Subscription successful for data streams %@", streams);
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf->_training1Btn setEnabled:true];
            [weakSelf->_training2Btn setEnabled:true];
        });
    };
    
    client.onSaveProfileOk = ^(NSString *profileName) {
        NSLog(@"Training profile saved %@", profileName);
    };
    
    client.onLoadProfileOk = ^(NSString *profileName) {
        NSLog(@"Training profile loaded %@",profileName);
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf->_training1Btn setEnabled:true];
            [weakSelf->_training2Btn setEnabled:true];
        });
    };
    
    client.onCreateProfileOk = ^(NSString *profileName) {
        NSLog(@"Training profile created %@", profileName);
    };

    
    client.onStreamDataReceived = ^(NSString *sessionId, NSString *stream, double time, NSArray *data) {
        if ([self->stream isEqual:@"facialExpression"] || [self->stream isEqual:@"mentalCommand"]) {
            if ([weakSelf isEvent: data event: @"Started"]) {
                NSLog(@"");
                NSLog(@"Please, focus on the action for a few seconds");
            } else if ([weakSelf isEvent: data event: @"Succeeded"]) {
                [weakSelf->client training: weakSelf->token sessionId: weakSelf->sessionId detection: weakSelf->stream action: weakSelf->action control: @"accept"];
            } else if ([weakSelf isEvent: data event: @"Failed"]) {
                dispatch_async(dispatch_get_main_queue(), ^{
                    [weakSelf enableButton:true];
                });
            } else if ([weakSelf isEvent: data event: @"Completed"]) {
                dispatch_async(dispatch_get_main_queue(), ^{
                    [weakSelf enableButton:true];
                });
                NSLog(@"Well done! You successfully trained %@", weakSelf->action);
            }
        }
        NSLog(@"\(%@) \(%@)", stream, data);
    };
    
    
    client.onTrainingOk = ^(NSString *msg) {
        // this signal is not important
        // instead we need to watch the events from the sys stream
    };
    
    client.onCreateRecordOk = ^(NSString *recordId){
        weakSelf->recordId = recordId;
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf->_stopRecordBtn setEnabled:true];
            [weakSelf->_injectMarkerBtn setEnabled:true];
        });
        NSLog(@"Create Record Successful");
    };
    
    client.onInjectMarkerOk = ^(NSString *markerId) {
        NSLog(@"Inject marker OK, marker id %@", (markerId));
        weakSelf->markerId = markerId;
    };
    
    client.onStopRecordOk = ^(NSString *recordId) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf->_stopRecordBtn setEnabled:false];
            [weakSelf->_injectMarkerBtn setEnabled:false];
        });
        [weakSelf->client getRecordInfos:weakSelf->token recordId:weakSelf->recordId];
        NSLog(@"Stop Record Successful");
    };
    
    client.onCreateSessionOk = ^(NSString *sessionId) {
        weakSelf->sessionId = sessionId;
        NSLog(@"Create Session Successful");
    };
    
    client.onCloseSessionOk = ^(){
        NSLog(@"Close session successful");
    };
}

- (void) handleHeadsetEvent {
    __unsafe_unretained typeof(self) weakSelf = self;
    self->finder = [[HeadsetFinder alloc] init];

    finder.onHeadsetFound = ^(Headset *headset, NSArray *headsets) {
        weakSelf->headset = headset;
        weakSelf->headsetList = headsets;
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf->_headsetTableView reloadData];
        });
    };
}

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    static NSString *cellIdentifier = @"cellID";
    
    HeadsetTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier forIndexPath:indexPath];
    if (cell == nil) {
        cell = [[HeadsetTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:cellIdentifier];
    }
    Headset* hs = headsetList[indexPath.row];
    cell.headsetName.text = hs->headsetId;
    cell.status.text = hs->status;
    cell.connectedBy.text = hs->connectedBy;
    return cell;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [self->headsetList count];
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 50;
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    return 50;
}

- (BOOL) isEvent: (NSArray *) data event: (NSString *) event  {
    for (NSString * val in data) {
        if ([val containsString:event]) {
            return true;
        }
    }
    return false;
}

- (void) enableButton: (BOOL) enable {
    [_training1Btn setEnabled:enable];
    [_training2Btn setEnabled:enable];
    [_createProfileBtn setEnabled:enable];
    [_loadProfileBtn setEnabled:enable];
    [_subscribeTrainingBtn setEnabled:enable];
}

@end

File: mobile/ios/CortexV2-Emotiv-iOS/CortexV2-Example-Swift/CortexV2-Example-Swift/DataStreamViewController.swift
//
//  DataStreamViewController.swift
//  CortexV2-Example-Swift
//
//  Created by Emotiv Inc on 2/21/20.
//  Copyright © 2020 Emotiv. All rights reserved.
//

import UIKit

class DataStreamViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {

    @IBOutlet weak var authorizeView: UIView!
    @IBOutlet weak var dataView: UIView!
    @IBOutlet weak var trainingView: UIView!
    @IBOutlet weak var headsetTableView: UITableView!
    @IBOutlet weak var markerView: UIView!
    @IBOutlet weak var streamView: UIView!
    
    @IBOutlet weak var subscribeBtn: UIButton!
    @IBOutlet weak var unsubscribeBtn: UIButton!
    @IBOutlet weak var createSession: UIButton!
    @IBOutlet weak var getLicenseInfo: UIButton!
    @IBOutlet weak var getUserInfo: UIButton!
    @IBOutlet weak var training1: UIButton!
    @IBOutlet weak var training2: UIButton!
    @IBOutlet weak var training: UIButton!
    @IBOutlet weak var loadProfile: UIButton!
    @IBOutlet weak var subscribeTraining: UIButton!
    @IBOutlet weak var createProfile: UIButton!
    @IBOutlet weak var injectMarker: UIButton!
    @IBOutlet weak var stopRecord: UIButton!
    @IBOutlet weak var createRecord: UIButton!
    @IBOutlet weak var warning: UILabel!
    
    var client = CortexClient.sharedInstance
    var creator = SessionCreator()
    var finder = HeadsetFinder()
    var headset = Headset()
    var headsetList: [Headset] = [Headset]()
    var activateSession: Bool = false
    var license: String = ""

    var token: String = ""
    var sessionId: String = ""
    var stream: String = ""
    
    var action: String = ""
    var recordId: String = ""
    var markerId: String = ""
    
    let cellReuseIdentifier = "HeadsetTableViewCell"
    
    override func viewDidLoad() {
        super.viewDidLoad()
        start()

        let headsetTableViewCell = UINib(nibName: cellReuseIdentifier, bundle: nil)
        self.headsetTableView.register(headsetTableViewCell, forCellReuseIdentifier: cellReuseIdentifier)
        headsetTableView.delegate = self
        headsetTableView.dataSource = self
        
        handleCortexEvent()
        handleHeadsetEvent()
        handleSessiontEvent()
        handleUI()
    }
    
    @IBAction func getUserLogin(_ sender: Any) {
        client.getUserLogin()
    }
    
    @IBAction func hasAccessRight(_ sender: Any) {
        client.hasAccessRight(clientId: ClientId, clientSecret: ClientSecret)
    }
    
    @IBAction func requestAccess(_ sender: Any) {
        client.requestAccess(clientId: ClientId, clientSecret: ClientSecret)
    }
    
    @IBAction func authorize(_ sender: Any) {
        client.authorize(clientId: ClientId, clientSecret: ClientSecret, license: license, debit: activateSession ? 1 : 0)
    }
    
    @IBAction func getUserInformation(_ sender: Any) {
        client.getUserInformation(token: token)
    }
    
    @IBAction func getLicenseInfo(_ sender: Any) {
        client.getLicenseInfos(token: token)
    }
    
    @IBAction func queryHeadsets(_ sender: Any) {
        finder.findHeadsets()
    }
    
    @IBAction func createSession(_ sender: Any) {
        creator.createSession(token: token, headset: headset, activate: activateSession, license: license)
    }
    
    @IBAction func subcribe(_ sender: Any) {
        client.subscribe(token: token, sessionId: sessionId, stream: stream)
    }
    
    @IBAction func unsubscribe(_ sender: Any) {
        client.unsubscribe(token: token, sessionId: sessionId, stream: stream)
    }
    
    @IBAction func subcribeTraining(_ sender: Any) {
        client.subscribe(token: token, sessionId: sessionId, stream: "sys")
    }
    
    @IBAction func createProfile(_ sender: Any) {
        client.createProfile(token: token, profileName: TrainingProfileName)
    }
    
    @IBAction func loadProfile(_ sender: Any) {
        client.loadProfile(token: token, headsetId: headset.id, profileName: TrainingProfileName)
    }
    
    @IBAction func trainNeutral(_ sender: Any) {
        self.action = "neutral"
        client.training(token: token, sessionId: sessionId, detection: stream , action: self.action, control: "start")
        self.enableButton(enable: false)
        NSLog("waiting training result from Cortex")
    }
    
    @IBAction func trainPush(_ sender: Any) {
        self.action = stream == "mentalCommand" ? "push" : "smile"
        client.training(token: token, sessionId: sessionId, detection: stream , action: self.action, control: "start")
        self.enableButton(enable: false)
        NSLog("waiting training result from Cortex")
    }
    
    @IBAction func createRecord(_ sender: Any) {
        client.createRecord(token: token, sessionId: sessionId, title: "Cortex Examples Swift")
    }
    
    @IBAction func injectMarker(_ sender: Any) {
        client.injectMarker(token: token, sessionId: sessionId, label: "test1", value: 41, time: Int64(Date().timeIntervalSince1970 * 1000));
    }
    
    @IBAction func stopRecord(_ sender: Any) {
        client.stopRecord(token: token, sessionId: sessionId)
    }
    
    @IBAction func training(_ sender: Any) {
        if sessionId == "" {
            NSLog("Please create session")
        } else {
            if self.stream == "session" {
                client.closeSession(token: self.token, sessionId: self.sessionId)
            } else {
                self.dataView.isHidden = true
                self.trainingView.isHidden = false
            }
        }
    }
    
    func start() {
        client.open()
    }
    
    func handleUI() {
        var nameStream = ""
        if stream == "mot" {
            nameStream = "motion"
        } else if stream == "eeg" {
            nameStream = "eeg"
        } else if stream == "pow" {
            nameStream = "Band Power"
        } else if stream == "com" {
            nameStream = "MC"
        } else if stream == "fac" {
            nameStream = "FE"
        } else if stream == "met" {
            nameStream = "PM"
        } else if stream == "facialExpression" {
            training.isHidden = false
            subscribeBtn.isHidden = true
            unsubscribeBtn.isHidden = true
            training1.setTitle("train FE Neutral", for: UIControl.State.normal)
            training2.setTitle("train FE Smile", for: UIControl.State.normal)
        } else if stream == "mentalCommand" {
            training.isHidden = false
            subscribeBtn.isHidden = true
            unsubscribeBtn.isHidden = true
            training1.setTitle("train MC Neutral", for: UIControl.State.normal)
            training2.setTitle("train MC Push", for: UIControl.State.normal)
        } else if stream == "marker" {
            streamView.isHidden = true
            markerView.isHidden = false
        }
        
        if stream != "authorize" {
            authorizeView.isHidden = true
            dataView.isHidden = false
        }

        if stream == "session" {
            training.isHidden = false
            subscribeBtn.isHidden = true
            unsubscribeBtn.isHidden = true
            self.training.setTitle("updateSession(Close)", for: UIControl.State.normal)
        }
        
        self.subscribeBtn.setTitle("subcribe " + nameStream, for: UIControl.State.normal)
        self.unsubscribeBtn.setTitle("unsubcribe " + nameStream, for: UIControl.State.normal)
        Timer.scheduledTimer(withTimeInterval: 60, repeats: false) { [weak self] (timer) in
            guard let weakSelf = self else { return }
            weakSelf.warning.isHidden = true
        }
    }
    
    func handleCortexEvent() {
        client.onConnected = { [weak self] in
        guard let weakSelf = self else { return }
            NSLog("Connected.")
            if weakSelf.stream != "authorize" {
                weakSelf.client.authorize(clientId: ClientId, clientSecret: ClientSecret, license: weakSelf.license, debit: weakSelf.activateSession ? 1 : 0)
            }
        }
        
        client.onDisconnected = {
            NSLog("Disconnected.")
        }
        
        client.onErrorReceived = { (method, code, message) in
            
        }
        
        client.onGetUserLoginOk = { (emotivId) in
           if emotivId == "" {
               NSLog("First, you must login with EMOTIV App")
               return
           }
           NSLog("You are logged in with the EmotivId \(emotivId)")
        }

        client.onUnsubscribeOk = { [weak self] (streams) in
            guard let weakSelf = self else { return }
            NSLog("Subscription cancelled for data streams \(streams)")
            weakSelf.client.closeSession(token: weakSelf.token, sessionId: weakSelf.sessionId)
        }
        
        client.onCloseSessionOk = {
            NSLog("Session closed.")
        }
        
        client.onGetLicenseInformationOk = {
            NSLog("Get License Information successful")
        }
        
        client.onGetUserInformationOk = {
            NSLog("Get User Information successful")
        }
        
        client.onAuthorizeOk = { [weak self] (token) in
            guard let weakSelf = self else { return }
            weakSelf.token = token
            NSLog("Authorize successful, token \(token)")
            if weakSelf.stream != "authorize" {
                weakSelf.finder.findHeadsets()
            }
        }
        
        client.onSubscribeOk = { [weak self] (streams) in
            guard let weakSelf = self else { return }
            NSLog("Subscription to data stream successful \(streams)")
            weakSelf.training1.isEnabled = true
            weakSelf.training2.isEnabled = true
        }
        
        client.onSaveProfileOk = { (profileName) in
            NSLog("Training profile saved \(profileName)")
        }
        
        client.onLoadProfileOk = { [weak self] (profileName) in
            guard let weakSelf = self else { return }
            NSLog("Training profile loaded \(profileName)")
            weakSelf.training1.isEnabled = true
            weakSelf.training2.isEnabled = true
        }
        
        client.onCreateProfileOk = { (profileName) in
            NSLog("Training profile created \(profileName)")
        }
        
        
        client.onStreamDataReceived = { [weak self] (sessionId, stream, time, data) in
            guard let weakSelf = self else { return }
            if weakSelf.stream == "facialExpression" || weakSelf.stream == "mentalCommand" {
                if weakSelf.isEvent(data: data, event: "Started") {
                    NSLog("")
                    NSLog("Please, focus on the action for a few seconds")
                } else if weakSelf.isEvent(data: data, event: "Succeeded") {
                    weakSelf.client.training(token: weakSelf.token, sessionId: weakSelf.sessionId, detection: weakSelf.stream, action: weakSelf.action, control: "accept")
                } else if weakSelf.isEvent(data: data, event: "Failed") {
                    weakSelf.enableButton(enable: true)
                } else if weakSelf.isEvent(data: data, event: "Completed") {
                    NSLog("Well done! You successfully trained")
                    weakSelf.enableButton(enable: true)
                }
            }
            NSLog("\(stream) \(data)")
        }
        
        client.onTrainingOk = { (msg) in
            // this signal is not important
            // instead we need to watch the events from the sys stream
        }
        
        client.onCreateRecordOk = { [weak self] (recordId) in
            guard let weakSelf = self else { return }
            weakSelf.recordId = recordId
            weakSelf.stopRecord.isEnabled = true
            weakSelf.injectMarker.isEnabled = true
        }
        
        client.onInjectMarkerOk = { [weak self] (markerId) in
            guard let weakSelf = self else { return }
            NSLog("Inject marker OK, marker id \(markerId)")
            weakSelf.markerId = markerId
        }
        
        client.onStopRecordOk = { [weak self] (recordId) in
            guard let weakSelf = self else { return }
            weakSelf.stopRecord.isEnabled = false
            weakSelf.injectMarker.isEnabled = false
            weakSelf.client.getRecordInfos(token: weakSelf.token, recordId: recordId)
        }
        
        client.onCloseSessionOk = {
            NSLog("Close session successful")
        }
    }
    
    private func handleHeadsetEvent() {
        finder.onHeadsetFound = { [weak self] (headset, headsets) in
            guard let weakSelf = self else { return }
            weakSelf.headset = headset
            weakSelf.headsetList = headsets
            weakSelf.headsetTableView.reloadData()
        }
    }
    
    func handleSessiontEvent() {
        creator.onSessionCreated = { [weak self] (token, sessionId) in
            guard let weakSelf = self else { return }
            weakSelf.token = token
            weakSelf.sessionId = sessionId
        }
    }
    
    private func enableButton(enable: Bool) {
        self.training1.isEnabled = enable
        self.training2.isEnabled = enable
        self.createProfile.isEnabled = enable
        self.loadProfile.isEnabled = enable
        self.subscribeTraining.isEnabled = enable
    }
    
    private func isEvent(data: NSArray, event: String) -> Bool {
        for val in data {
            if val is String && (val as! String).contains(event) {
                return true
            }
        }
        return false
    }
    
    // number of rows in table view
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return headsetList.count
    }

    // create a cell for each table view row
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let hscell = tableView.dequeueReusableCell(withIdentifier: cellReuseIdentifier) as? HeadsetTableViewCell
        hscell?.status.text = headsetList[indexPath.row].status
        hscell?.headsetName.text = headsetList[indexPath.row].id
        hscell?.connectedBy.text = headsetList[indexPath.row].connectedBy
        return hscell ?? UITableViewCell()
    }
    
    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        let headerCell = tableView.dequeueReusableCell(withIdentifier: cellReuseIdentifier) as? HeadsetTableViewCell
        return headerCell ?? UITableViewCell()
    }
    
    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 50
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 50
    }
}

